### PR TITLE
Show captions of image in media details

### LINKED ISSUE #2337: Show caption in media details
from Structured Commons.
Show on first line, before title, in user's language or fallback language.

### PR DESCRIPTION
**Pull request by sp2710**
**Description (required)**

Fixes #2337

**What changes did you make and why?**

- fetched captions of media by using mediawiki api
- Added captions section in media details
- Added unit test cases

**Tests performed (required)**

Tested ProdDebug on Motorola Moto G5 plus with API level 24.

**Screenshots showing what changed (optional - for UI changes)**

device-2019-03-19-130232

### CODE REVIEW

**Comment by sp2710** [comment]
@nicolas-raoul I have used english as the default language for testing and also have enabled language fallback .
What should be the default language in which captions should be shown?

**Comment by nicolas-raoul** [comment]
Looks great!

Please use the Android locale. For instance, if the Android phone is in French, the caption should be shown in French.
Multilingualism is the goal of captions, actually :-)
I just added French captions to https://commons.wikimedia.org/wiki/File:M%C3%BCnster,_St.-Paulus-Dom,_Alter_Chor_--_2019_--_3789-93.jpg

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -172,6 +173,7 @@
   <string name="upload_image_duplicate">This file already exists on Commons. Are you sure you want to proceed?</string>
   <string name="yes">Yes</string>
   <string name="no">No</string>
+  <string name="media_detail_captions">Captions</string>
```
Comment (nicolas-raoul) [inline app/src/main/res/values/strings.xml]: We only display one caption, so singular is better than plural :-)

**Comment by sp2710** [comment]
@nicolas-raoul I have updated my PR

File: app/src/main/java/fr/free/nrw/commons/Media.java [hunk 2]
```diff
@@ -180,6 +181,22 @@ public String getImageUrl() {
         return imageUrl;
     }
 
+    /**
+     * Sets the Captions of the file.
+     * @param captions
+     */
+    public void setCaptions(String captions) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.java]: plural -> singular

File: app/src/main/java/fr/free/nrw/commons/Media.java [hunk 3]
```diff
@@ -180,6 +181,22 @@ public String getImageUrl() {
...
+        this.captions = captions;
+    }
+
+    /**
+     * Gets the file Captions as a string.
+     * @return file Captions as a string
+     */
+    public String getCaptions() {
+        return captions;
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.java]: singular

File: app/src/main/java/fr/free/nrw/commons/MediaDataExtractor.java [hunk 4]
```diff
@@ -40,6 +40,7 @@
     private boolean deletionStatus;
     private ArrayList<String> categories;
     private Map<String, String> descriptions;
+    private String captions;
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/MediaDataExtractor.java]: singular

File: app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java [hunk 5]
```diff
@@ -317,6 +317,19 @@ public String parseWikicode(String source) throws IOException {
                 .getString("/api/flow-parsoid-utils/@content");
     }
 
+    @Override
+    public String fetchCaptionsByFilename(String filename) throws IOException {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java]: singular
javadoc

File: app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java [hunk 6]
```diff
@@ -317,6 +317,19 @@ public String parseWikicode(String source) throws IOException {
...
+        return api.action("wbgetentities")
+                .param("sites", "commonswiki")
+                .param("titles", filename)
+                .param("props", "labels")
+                .param("format", "xml")
+                .param("languages", Locale.getDefault().getLanguage())
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java]: Looks OK! :-)

**Comment by sp2710** [comment]
@nicolas-raoul I have made it singular and have added java docs :-)

**Comment by sp2710** [comment]
@maskaravivek Sure I will add it :)

**Comment by sp2710** [comment]
@maskaravivek I have added unit test cases.

**Comment by neslihanturan** [comment]
I am closing this one since another one is already available for this issue, see #2933

**Pull request closed without merging**